### DISCUSSION TITLE
Amalgamtion of wesen1/LuaORM

### DISCUSSION THREAD

**Discussion by erichowey** [siffiejoe/lua-amalg#11] [Q&A] [answered]
Hi There,

Sorry for the noobie question. I'm really enjoying this application so far!

I'm trying to run amalgamation on the follow repo into my project.
https://github.com/wesen1/LuaORM

I have LuaORM installed into my system via luarocks. My amalg.cache file looks something like this:

https://gist.github.com/erichowey/50a9a29568081e9188f1f1181a8d6cdd

However, I can't seem to get the directory structure to work right. For example: 
`/usr/local/share/lua/5.3/LuaORM/LuaORM/ORM.lua` exists on my machine, however, amalgamtion outputs something like this:
```lua
  -- API to configure the ORM
  local ORM = require("LuaORM/ORM")
  LuaORM_API.ORM = ORM(sourceDirectoryPath)
```

Which in turns throws the following error on my application:
```
...
[string "extensions.lua"]:36: module 'LuaORM/ORM' not found:
	no field package.preload['LuaORM/ORM']
	no file '/usr/share/lua/5.3/LuaORM/ORM.lua'
	no file '/usr/share/lua/5.3/LuaORM/ORM/init.lua'
	no file '/usr/lib64/lua/5.3/LuaORM/ORM.lua'
	no file '/usr/lib64/lua/5.3/LuaORM/ORM/init.lua'
	no file './LuaORM/ORM.lua'
	no file './LuaORM/ORM/init.lua'
	no file '[string "extensions.lua"]/LuaORM/ORM.lua'
	no file '/usr/lib64/lua/5.3/LuaORM/ORM.so'
	no file '/usr/lib64/lua/5.3/loadall.so'
	no file './LuaORM/ORM.so'
...
```

When believe the module should actually be:
```
LuaORM/LuaORM/ORM
```

Any advice would be greatly appreciated. Thank You!

**Comment by siffiejoe** [comment, marked as answer]
Hi!

This will be a tough one, because LuaORM very much depends on file paths derived from Lua module files.

Your `amalg.cache` file looks good, but it doesn't match with how LuaORM uses its modules. How did you generate it? By hand?!

> [string "extensions.lua"]:36: module 'LuaORM/ORM' not found:

Basically, the module name (key in the cache) must be exactly the same as the argument to `require` in the source code of LuaORM, and the same as the module name in the error message. If LuaORM wants `LuaORM/ORM`, this is what must be in the cache, not `LuaORM.ORM` or even `LuaORM.LuaORM.ORM`. Most modules in the LuaORM source code are `require`d this way. (For the record: you should not use directory separators in module names!)

The second issue is that LuaORM uses the `debug` module to figure out the file path of the first module and changes `package.path` accordingly. This is why `require("LuaORM/ORM")` works at all, and this change is what we must recreate during amalgamation.

This should be possible using something like this:
```sh
LUA_PATH_5_3="/usr/share/lua/5.3/LuaORM/?.lua;;" amalg.lua -s script.lua -o out.lua -c
```
And the third issue is that LuaORM uses the path derived from the first module as the base path for reading template files. The derived path is probably nonsense because the first module doesn't exist as a file anymore after amalgamation but `amalg.lua` does not embed non-module files anyway. You can try adding the debug switch (`-d`) to maybe get a sensible path from where to read the templates, but the templates will definitely not be in the amalgamated output file.

HTH,
Philipp

**Reply by erichowey** [reply]
Thank you so much for the response! This is wonderful.

> Your amalg.cache file looks good, but it doesn't match with how LuaORM uses its modules. How did you generate it? By hand?!

Yes. I used `lua -lamalg` to generate it at first, but then `amalg.lua` failed afterwards due to errors in the cache file, specifically, the preceding `LuaORM` directory missing from all of the entries. I modified the cache file until `amalg.lua` ran successfully. `lua -lamalg` outputted directory separators as an FYI.

Thank you very much for the other tips. I'll will work with this and post updates.

**Reply by siffiejoe** [reply]
> I modified the cache file until `amalg.lua` ran successfully. `lua -lamalg` outputted directory separators as an FYI.

I know. `lua -lamalg` outputs the module names exactly like they are used by the application (even with directory separators if necessary), therefore you should always tweak `package.path` instead of "fixing" the cache if a module isn't found during amalgamation.

This is probably something I should mention in the `README.md` somewhere even if programs that modify their own `package.path` are rare.

**Reply by erichowey** [reply]
After toying with this over the past few days, I kept running into namespace conflicts with LuaORM after correcting the path of the application. I think I'm going to give up on it and try to find another approach for this project. It's a shame because LuaORM solves a major headache for me even as an immature library.

Thank you so much for taking the time to look into this. I really appreciate it!
